src/mlpack/methods/kde/kde_main.cpp

```diff
@@ -0,0 +1,165 @@
+/**
+ * @file kde_main.cpp
+ * @author Roberto Hueso (robertohueso96@gmail.com)
+ *
+ * Executable for running Kernel Density Estimation.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/core/util/cli.hpp>
+#include <mlpack/core/util/mlpack_main.hpp>
+#include <mlpack/core/tree/statistic.hpp>
+
+#include "kde.hpp"
+#include "kde_model.hpp"
+
+using namespace mlpack;
+using namespace mlpack::kde;
+using namespace mlpack::util;
+using namespace std;
+
+// Define parameters for the executable.
+PROGRAM_INFO("Kernel Density Estimation",
+    "This program performs a Kernel Density Estimation. KDE is a "
+    "non-parametric way of estimating probability density function. "
+    "For each query point the program will estimate its probability density "
+    "by applying a kernel function to each reference point. Computational "
+    " complexity is O(n^2) but it is optimized by making use of dual-trees. "
+    "\n\n"
+    "For example, the following will run KDE using the points in "
+    "reference_set.csv and query_set.csv. It will apply an Epanechnikov kernel "
+    "with a 0.2 bandwidth to each reference point and use a KD-Tree for the "
+    "dual-tree optimization. The result will be stored in a densities.csv file "
+    "with a maximum error of 5%"
+    "\n\n"
+    "$ kde --reference reference_set.csv --query query_set.csv --bandwidth 0.2 "
+    "--kernel epanechnikov --tree kd-tree --rel_error 0.05 --output "
+    "densities.csv"
+    "\n\n"
+    "Dual-tree optimization allows to avoid lots of barely relevant "
+    "calculations (as kernel function values decrease with distance) if you "
+    "can afford a little error (you can define how much is the maximum you are "
+    "willing to afford) over the final result. This program runs using an "
+    "Euclidean metric. If no output file is specified then it will output the "
+    "result to standard output.");
+
+// Required options.
+PARAM_MATRIX_IN("reference", "Input dataset to KDE on.", "r");
+PARAM_MATRIX_IN_REQ("query", "Query dataset to KDE on.", "q");
+PARAM_DOUBLE_IN("bandwidth", "Bandwidth of the kernel", "b", 1.0);
+
+// Load or save models.
+PARAM_MODEL_IN(KDEModel,
+               "input_model",
+               "File containing pre-trained KDE model.",
+               "m");
+PARAM_MODEL_OUT(KDEModel,
+                "output_model",
+                "If specified, the KDE model will be saved to the given file.",
+                "M");
+
+// Configuration options
+PARAM_STRING_IN("kernel", "Kernel to use for the estimation"
+    "('gaussian', 'epanechnikov').", "k", "gaussian");
+PARAM_STRING_IN("tree", "Tree to use for the estimation"
+    "('kd-tree', 'ball-tree').", "t", "kd-tree");
```

Comment:
Up to you, maybe we could add other tree types?  Maybe that causes problems with the breadth-first traversal.

Reply:
Added Cover-tree, Octree and R-Tree in 9cff9c5b6d7ffc44d8b764652a808d4bda798bfc.
I took the decision not to include breadth-first traversal for the moment since you can do that from the C++ API.
I mark this as solved but I will probably add more trees in the future :smile: